controllers/etcd_controller.go

```diff
@@ -1123,11 +1208,142 @@ func (r *EtcdReconciler) waitUntilStatefulSetReady(sts *appsv1.StatefulSet) (*ap
 			return gardenerretry.MinorError(err)
 		}
+		if ss.Status.Replicas != *ss.Spec.Replicas {
+			return gardenerretry.MinorError(fmt.Errorf("statefulset not scaled"))
+		}
 		return gardenerretry.Ok()
 	})
 	return ss, err
 }
 
+func (r *EtcdReconciler) getStatefulSet(ctx context.Context, etcd *druidv1alpha1.Etcd) (*appsv1.StatefulSet, error) {
+	selector, err := metav1.LabelSelectorAsSelector(etcd.Spec.Selector)
+	if err != nil {
+		logger.Error(err, "error converting etcd selector to labelSelector")
+		return nil, err
+	}
+
+	statefulSets := &appsv1.StatefulSetList{}
+	if err = r.List(ctx, statefulSets, client.InNamespace(etcd.Namespace), client.MatchingLabelsSelector{Selector: selector}); err != nil {
+		return nil, err
+	}
+
+	if len(statefulSets.Items) > 0 {
+		return &statefulSets.Items[0], nil
+	}
+	return nil, nil
+}
+
+func (r *EtcdReconciler) getService(ctx context.Context, etcd *druidv1alpha1.Etcd) (*v1.Service, error) {
+	selector, err := metav1.LabelSelectorAsSelector(etcd.Spec.Selector)
+	if err != nil {
+		logger.Error(err, "error converting etcd selector to labelSelector")
+		return nil, err
+	}
+
+	services := &corev1.ServiceList{}
+	if err := r.List(ctx, services, client.InNamespace(etcd.Namespace), client.MatchingLabelsSelector{Selector: selector}); err != nil {
+		logger.Error(err, "error listing services")
+		return nil, err
+	}
+
+	if len(services.Items) > 0 {
+		return &services.Items[0], nil
+	}
+	return nil, nil
+}
+
+func (r *EtcdReconciler) listPersistentVolumeClaims(ss *appsv1.StatefulSet) (*corev1.PersistentVolumeClaimList, error) {
+	selector, err := metav1.LabelSelectorAsSelector(ss.Spec.Selector)
+	if err != nil {
+		logger.Error(err, "error converting statefulset selector to selector")
+		return nil, err
+	}
+
+	persistentVolumeClaimList := &corev1.PersistentVolumeClaimList{}
+	if err := r.List(context.TODO(), persistentVolumeClaimList, client.InNamespace(ss.Namespace), client.MatchingLabelsSelector{Selector: selector}); err != nil {
+		return nil, err
+	}
+
+	return persistentVolumeClaimList, nil
+}
+
+func (r *EtcdReconciler) prepareEtcdForScaleDown(parentCtx context.Context, etcd *druidv1alpha1.Etcd, ss *appsv1.StatefulSet, svc *v1.Service) (*corev1.PersistentVolumeClaimList, error) {
+	// TODO: wait for etcd health to be false so that etcd is not serving traffic
+
+	if parentCtx == nil {
+		parentCtx = context.TODO()
+	}
+	ctx, cancel := context.WithTimeout(parentCtx, DefaultTimeout)
+	defer cancel()
+	if err := r.takeFullSnapshot(ctx, etcd, svc); err != nil {
+		return nil, err
+	}
+
+	pvcList, err := r.listPersistentVolumeClaims(ss)
+	if err != nil {
+		return nil, err
+	}
+
+	return pvcList, nil
+}
+
+func (r *EtcdReconciler) removePVCs(pvcList []corev1.PersistentVolumeClaim) error {
+	resultCh := make(chan utils.PVCDeletionResult, len(pvcList))
+
+	for _, pvc := range pvcList {
+		go func(resultCh chan<- utils.PVCDeletionResult, pvc v1.PersistentVolumeClaim) {
```

Comment:
I didn't mean that we should delete the PVCs in goroutines. Just that we should fire the deletes and aggregate the errors in a loop and then outside the loop check only once if there are errors and _enqueueAfter_.